Feature lb cache

drmaas opened the pull request.

Commit:
adding load balancer caching with relationships to server groups and instances

Commit:
load balancer caching complete and tested

lwander committed:
provider/google: fix loading of lb tab (#776)

danielpeach committed:
(gce) added upsert google scaling policy atomic operation (#743)

varikin commented on clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/Keys.groovy:
```diff
@@ -88,12 +91,12 @@ class Keys {
           case Namespace.SUBNETS.ns:
             if (parts.length == 5) {
-              result << [id: parts[2], account: parts[3], region: parts[4]]
+              result << [account: parts[2], region: parts[3], id: parts[4]]
```
Why re-arrange this?

drmaas replied:
It was not consistent with other keys.

varikin commented on clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/Keys.groovy:
```diff
@@ -155,6 +183,18 @@ class Keys {
   }
 
+  static String getVipKey(String vipId, String account, String region) {
+    "${ID}:${Namespace.VIPS}:${account}:${region}:${vipId}"
```
The keys above are parsing id, account region, but this is doing account, region, id. I don't think that will work.

drmaas replied:
It does work - I made it consistent across keys.

varikin replied:
From above:

```
case Namespace.VIPS.ns:
        if (parts.length == 5) {
           result << [id: parts[2], account: parts[3], region: parts[4]]
  }
```

When parsing the VIP key, you get id, account, region. When creating the key, you do account, region, id. I don't know how that will work.

drmaas replied:
I checked and all are consistent and working. 

Here's a quick test:

```
groovy> println "1:2:3".split(":") 

[1, 2, 3]
groovy> def parts = "1:2:3".split(":") 
groovy> println parts[2] 

3
```

You can see it splits left to right.

varikin commented on clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/controllers/OpenstackLoadBalancerController.groovy:
```diff
@@ -0,0 +1,56 @@
+
+  @RequestMapping(method = RequestMethod.GET)
+  Set<OpenstackLoadBalancerSummary> list() {
+    provider.getApplicationLoadBalancers('').collect { lb ->
```
What is `''` in the call to get load balancers?

drmaas replied:
It is the application name, which is required to  adhere to the provider interface. We pass `''` because the app name doesn't help us find load balancers, as they are keyed off account and region.

lwander committed:
provider/consul: Refactored discovery health (#778)

varikin commented on clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackFloatingIP.groovy:
```diff
@@ -0,0 +1,42 @@
+   *
+   * @param vip
+   * @return
+   */
```
Please fill out the javadoc or remove it

varikin commented on clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackLoadBalancer.groovy:
```diff
@@ -16,17 +16,45 @@
+
+  static OpenstackLoadBalancer from(LbPool pool, OpenstackVip vip, OpenstackSubnet subnet, OpenstackFloatingIP ip,
+                                    Set<HealthMonitor> healthMonitors, String account, String region) {
+    new OpenstackLoadBalancer(account: account, region: region, id: pool?.id, name: pool?.name, description: pool?.description,
```
It seems weird that `pool` could be null. Would it be better to have that be an error?

drmaas replied:
It will never be null, so yes.

varikin commented on clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackLoadBalancerSummary.groovy:
```diff
@@ -0,0 +1,32 @@
+
+/**
+ *
+ */
```
Remove empty javadoc

varikin commented on clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackPort.groovy:
```diff
@@ -0,0 +1,40 @@
+   *
+   * @param port
+   * @return
+     */
```
Remove empty javadoc

varikin commented on clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackServerGroup.groovy:
```diff
@@ -45,38 +46,44 @@ class OpenstackServerGroup implements ServerGroup, Serializable {
   @Override
   Capacity getCapacity() {
     scalingConfig ?
       new Capacity(min: scalingConfig.minSize ? scalingConfig.minSize as Integer : 0, max: scalingConfig.maxSize ? scalingConfig.maxSize as Integer : 0)
```
what is `maxSize` and `minSize` that you are casting to integer?

drmaas replied:
@derekolk created scalingConfig as `Map<String, Object> scalingConfig`. I didn't change it.

varikin commented on clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackVip.groovy:
```diff
@@ -0,0 +1,40 @@
+   * @param vip
+   * @return
+   */
+  static OpenstackVip from(Vip vip, String account, String region) {
```
Remove empty javadoc

varikin committed:
Added remote security group to security group description. (#777)

Allows upsertting an Openstack security group that references another
security group.

Also, including handling of ICMP rules for openstack security groups.

lwander committed:
provider/consul: Provide a way to get default config (#779)

Commit:
improve Keys test converage, address PR comments

drmaas commented:
Closing due to message PR history.
